File: back/test/cat/feeder.py
#!/usr/bin/python

from contextlib import contextmanager
import os
import string
import sys
import termios


@contextmanager
def raw_mode(fd, when=termios.TCSAFLUSH):
    """Put terminal into a raw mode."""
    # Indexes for termios list.
    IFLAG = 0
    OFLAG = 1
    CFLAG = 2
    LFLAG = 3
    ISPEED = 4
    OSPEED = 5
    CC = 6
    saved_mode = termios.tcgetattr(fd)
    raw_mode = saved_mode[:]

    cc = raw_mode[CC]
    print 'CC', cc
    print 'VSTOP %s' % readable(cc[termios.VSTOP])
    print 'VSTART %s' % readable(cc[termios.VSTART])
    print 'CRTSCTS %#x' % (raw_mode[CFLAG] & termios.CRTSCTS)

    raw_mode[IFLAG] &= ~(termios.BRKINT | termios.ICRNL |
                         termios.INPCK | termios.ISTRIP)
  # raw_mode[IFLAG] |=  (termios.IXON | termios.IXOFF) # XXX kbob
  # raw_mode[IFLAG] |=   termios.IXON # XXX kbob
    raw_mode[IFLAG] &= ~(termios.IXON) # XXX kbob
    raw_mode[OFLAG] &= ~(termios.OPOST)
    raw_mode[CFLAG] &= ~(termios.CSIZE | termios.PARENB)
    raw_mode[CFLAG] |=  termios.CS8
#   raw_mode[CFLAG] &= ~(termios.CRTSCTS) # XXX kbob
    raw_mode[LFLAG] &= ~(termios.ECHO | termios.ICANON |
                         termios.IEXTEN | termios.ISIG)
    raw_mode[CC][termios.VMIN]  = 1
    raw_mode[CC][termios.VTIME] = 1
    termios.tcsetattr(fd, when, raw_mode)
    try:
        yield 
    finally:
        termios.tcsetattr(fd, when, saved_mode)

@contextmanager
def os_open(file, name, mode=0666):
    fd = os.open(file, name, mode)
    try:
        yield fd
    finally:
        os.close(fd)

def readable(c):
    i = ord(c)
    if 32 <= i < 0177 or c in '\n':
        return c
    return '\033[1m^%c\033[m' % chr(i ^ 0100)

def accum_input(ttyfd):
    data = ''
    while True:
        try:
            datum = os.read(ttyfd, 1)
        except OSError, x:
            print >>sys.stdout, x
            continue
        os.write(sys.stdout.fileno(), readable(datum))
        data += datum
        yield data

def test_pattern():
    w = [c * 4 for c in string.lowercase]
    lines = [' '.join(w[i:i+4]) + '\r\n' for i in range(0, len(w), 4)]
    return ''.join(lines)

def test_cat():
    print '[feeder]'
    sent = False
    with os_open('/dev/ttyUSB0', os.O_RDWR) as ttyfd, raw_mode(ttyfd):
        for data in accum_input(ttyfd):
            if not sent and 'Ready\r\n' in data:
                print 'Found ready prompt'
                os.write(ttyfd, test_pattern() * 2)
                sent = True

if __name__ == '__main__':
    try:
        test_cat()
    except KeyboardInterrupt:
        print
        exit(1)

File: back/test/cat/feeder.c
#include <assert.h>
#include <stdbool.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <termios.h>
#include <unistd.h>
#include <sys/fcntl.h>
#include <sys/select.h>
#include <sys/stat.h>

#define BACK_RX_BUF_SIZE 256
#define FLOW_SHIFT 4

struct termios orig_tios, raw_tios;
char     buf0[BUFSIZ];
size_t   count0;
uint8_t  tx_mark;
uint16_t tx_received;
uint16_t tx_limit;
uint16_t tx_count;

char     buf1[BUFSIZ];
size_t   count1;

void pstat(int fd, const char *label)
{
    struct stat s;
    if (fstat(fd, &s) < 0)
        perror(label);
    else
        printf("%s: blksize = %ld\n", label, (long)s.st_blksize);
}

int putc_readable(char c, FILE *f)
{
    char c1 = c;
    if ((' ' <= c && c < '\177') || c == '\n')
#if 1
        return putc(c1, f);
#else
        return 0;
#endif
    const char *mp = "";
    if ((uint8_t)c1 >= 0xF0) {
        return fprintf(f, "\33[1mM%x\33[m", c1 & 0x0F);
    } else if (c1 & 0200) {
        c1 &= ~0200;
        mp = "M-";
        if (' ' <= c1 && c1 < '\177')
            return fprintf(f, "\033[1mM-%c\033[m", c1) == EOF ? EOF : c;
    }
    return fprintf(f, "\033[1m%s^%c\033[m", mp, c1 ^ 0100) == EOF ? EOF : c;
}

void calc_limit(uint8_t mark)
{
    uint8_t lo = tx_received & 0x00FF;
    uint16_t hi = tx_received & 0xFF00;
    if (mark < lo)
        hi += 0x0100;
    // printf("calc tx_received %#04x -> %#04x, tx_limit %#04x -> %#04x\n",
    //        tx_received, (hi | mark), tx_limit, (hi | mark) + 0xFF);
    fflush(stdout);
    tx_received = hi | mark;
    tx_limit = tx_received + 0xFF;
}

uint8_t tx_space(void)
{
    return tx_limit - tx_count;
}

int main()
{
    int infd = fileno(stdin);
    int outfd = fileno(stdout);
    int ttyfd = open("/dev/ttyUSB0", O_RDWR | O_NONBLOCK | O_NOCTTY, 0666);
    if (ttyfd < 0)
        perror("/dev/ttyUSB0"), exit(EXIT_FAILURE);

    pstat(infd, "standard input");
    pstat(outfd, "standard output");
    pstat(ttyfd, "serial");

    if (tcgetattr(ttyfd, &orig_tios) != 0)
        perror("tcgetattr"), exit(EXIT_FAILURE);
    raw_tios = orig_tios;
    raw_tios.c_iflag &= ~(IGNBRK | BRKINT | PARMRK | ISTRIP |
                          INLCR | IGNCR | ICRNL | IXON);
    raw_tios.c_oflag &= ~ OPOST;
    raw_tios.c_lflag &= ~(ECHO | ECHONL | ICANON | ISIG | IEXTEN);
    raw_tios.c_cflag &= ~(CSIZE | PARENB);
    raw_tios.c_cflag |=   CS8;
    if (tcsetattr(ttyfd, TCSAFLUSH, &raw_tios) != 0)
        perror("tcsetattr"), exit(EXIT_FAILURE);

    while (true) {
        int nfds = ttyfd + 1;;
        fd_set readfds, writefds;
        fd_set *writefdp = NULL;
        FD_ZERO(&readfds);
        FD_ZERO(&writefds);
        FD_SET(ttyfd, &readfds);
        if (count0 && tx_space()) {
            writefdp = &writefds;
            FD_SET(ttyfd, writefdp);
        } else {
            FD_SET(infd, &readfds);
            if (infd >= nfds)
                nfds = infd + 1;
        }
        int r_nfds = select(nfds, &readfds, writefdp, NULL, NULL);
        if (r_nfds < 0)
            perror("select"), exit(EXIT_FAILURE);

        if (FD_ISSET(ttyfd, &readfds)) {

            // Read from serial (Highest priority)
            ssize_t nr = read(ttyfd, buf1, sizeof buf1);
            if (nr == -1)
                perror("read tty"), exit(EXIT_FAILURE);
            else {
                size_t i;
                for (i = 0; i < nr; i++) {
                    char c = buf1[i];
                    uint8_t uc = (uint8_t)c;
                    if ((uc & 0xF0) == 0xF0) {
                        tx_mark = (uc & ~-(1 << FLOW_SHIFT)) << FLOW_SHIFT;
                        calc_limit(tx_mark);
                    } else {
                        putc_readable(c, stdout);
                        fflush(stdout);
                    }
                }
            }
        } else if (!count0 && FD_ISSET(infd, &readfds)) {

            // Read from stdin.
            ssize_t nr = read(infd, buf0, sizeof buf0);
            if (nr < 0)
                perror("read stdin"), exit(EXIT_FAILURE);
            count0 = nr;
            // printf("read %zd; count0=%zd\n", nr, count0);
        }
        else if (count0 && tx_space() && FD_ISSET(ttyfd, writefdp)) {

            // Write to serial.
            size_t ntw = tx_space();
            if (ntw > count0)
                ntw = count0;
            ssize_t nw = write(ttyfd, buf0, ntw);
            if (nw < 0)
                perror("write"), exit(EXIT_FAILURE);
            else {
                if (nw && nw < count0)
                    memmove(buf0, buf0 + nw, count0 - nw);
                count0 -= nw;
                tx_count += nw;
                // printf("writ %zd; count0=%zd\n", nw, count0);
            }
        }
    }

    if (tcsetattr(ttyfd, TCSAFLUSH, &orig_tios) != 0)
        perror("tcsetattr"), exit(EXIT_FAILURE);
    return EXIT_SUCCESS;
}
